packages/vite/src/node/server/hmr.ts

```diff
@@ -77,7 +78,22 @@ export async function handleHMRUpdate(
       { clear: true, timestamp: true },
     )
     try {
+      const { port: prevPort, host: prevHost, middlewareMode } = config.server
+      const oldUrls = server.resolvedUrls
       await server.restart()
+      const {
+        logger,
+        server: { port, host },
+      } = server.config
+      if (
+        !middlewareMode &&
+        ((port ?? DEFAULT_DEV_PORT) !== (prevPort ?? DEFAULT_DEV_PORT) ||
+          host !== prevHost ||
+          diffDnsOrderChange(oldUrls, server.resolvedUrls))
+      ) {
+        logger.info('')
+        server.printUrls()
+      }
```

Comment:
I think we need to do this in `packages/vite/src/node/shortcuts.ts` as well.

Reply:
That is a bit tricky though since if they call `createServer` programmatically, and call `server.bindCLIShortcuts`, then the `r + enter` shortcut would also print Vite's urls which might not be what they want.

Maybe we can take a simpler approach to this, e.g. if `printUrls` is called, we mark `server._printedUrls: true`, and on restart check against the variable?

Reply:
We already have `u + enter` shortcut that prints Vite's urls. Should this shortcut be disabled in that case?
https://github.com/vitejs/vite/blob/cc9fb8736cf7fd974c50b98ea9ecf9d8bb61b22f/packages/vite/src/node/shortcuts.ts#L97-L104

> Maybe we can take a simpler approach to this, e.g. if `printUrls` is called, we mark `server._printedUrls: true`, and on restart check against the variable?

I think that would work. But people won't expect calling `printUrls` will have a side effect (setting `server._printedUrls: true`) and might be confused when they encountered that behavior.

Reply:
I was thinking more of "if `printUrls` is never called, then `r + enter` also won't print the url", and the side effect should never really be seen in practice.

But yeah in any case, I guess we could create an internal `server._restart` API to abstract the default restart logic here. Maybe we should think about how users can override shortcuts so they re-implement the shortcuts or remove it themselves, before Vite 5.

Reply:
Nevermind it looks like users can define a stub shortcut to "disable" it